compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -285,20 +285,33 @@ pub trait Machine<'mir, 'tcx>: Sized {
         ptr: Pointer,
     ) -> Pointer<Self::PointerTag>;
 
-    /// "Int-to-pointer cast"
+    /// Hook for returning a pointer from a transmute-like operation on an addr.
     fn ptr_from_addr(
         ecx: &InterpCx<'mir, 'tcx, Self>,
         addr: u64,
     ) -> Pointer<Option<Self::PointerTag>>;
 
+    /// "Int-to-pointer cast"
+    fn ptr_from_casted_addr(
+        ecx: &InterpCx<'mir, 'tcx, Self>,
+        addr: u64,
+    ) -> Pointer<Option<Self::PointerTag>>;
+
+    /// Marks a pointer as exposed, allowing it's provenance
+    /// to be recovered. "Pointer-to-int cast"
+    fn expose_ptr(
+        ecx: &mut InterpCx<'mir, 'tcx, Self>,
+        ptr: Pointer<Self::PointerTag>,
+    ) -> InterpResult<'tcx>;
+
     /// Convert a pointer with provenance into an allocation-offset pair
     /// and extra provenance info.
     ///
     /// The returned `AllocId` must be the same as `ptr.provenance.get_alloc_id()`.
     fn ptr_get_alloc(
```

Comment:
Please explain in the doc comment when this is allowed to fail and what that means.

Reply:
I'm not sure how to word the doc comment, but I've put in the general idea in 
170e2ef.

Reply:
I was talking about ptr_get_alloc, not expose_addr. :)